'use strict';

// MODULES //

var tape = require( 'tape' );
var noop = require( '@stdlib/utils/noop' );
var wrap = require( './../lib' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof wrap, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function will throw an error if not provided a function', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		5,
		null,
		NaN,
		undefined,
		[],
		{}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			wrap( value );
		};
	}
});

tape( 'the function returns a function', function test( t ) {
	t.strictEqual( typeof wrap( noop ), 'function', 'returns a function' );
	t.end();
});

tape( 'the function returns a function having the same signature', function test( t ) {
	var f;

	f = wrap( add );
	t.strictEqual( add( 1, 2, 3 ), f( 1, 2, 3 ), 'returns same results' );
	t.end();

	function add( a, b, c ) {
		return a + b + c;
	}
});

tape( 'the function returns any trapped errors', function test( t ) {
	var out;
	var f;

	f = wrap( fcn );
	out = f();

	t.strictEqual( out instanceof Error, true, 'returns an error instance' );
	t.strictEqual( out.message, 'beep boop', 'returns trapped error' );
	t.end();

	function fcn() {
		throw new Error( 'beep boop' );
	}
});

tape( 'the function supports setting the `this` context', function test( t ) {
	var f;

	f = wrap( fcn, {
		'beep': 'boop'
	});

	t.strictEqual( f(), 'boop', 'returns property bound to `this` context' );
	t.end();

	function fcn() {
		return this.beep; // eslint-disable-line no-invalid-this
	}
});

tape( 'the function will not trap and return errors which occur in nested asynchronous function scopes', function test( t ) {
	var f = wrap( fcn );
	t.strictEqual( f(), void 0, 'returns undefined' );

	function fcn() {
		setTimeout( clbk, 0 );
	}
	function clbk() {
		t.throws( createError, Error, 'throws an untrapped error' );
		t.end();
	}
	function createError() {
		throw new Error( 'beep boop' );
	}
});
